src/lsm/manifest_level.zig

```diff
@@ -54,6 +54,101 @@ pub fn ManifestLevelType(
+            /// contributes to the level's key_min/key_max, we find a new key_min/key_max.
+            ///
+            /// This is achieved by querying the tables visible to snapshot_latest and updating level
+            /// key_min/key_max to the key_min/key_max of the first table returned by the iterator.
+            /// The query is guaranteed to only fetch non-snapshotted tables, since
+            /// tables visible to old snapshots that users have retained would have
+            /// snapshot_max set to a non math.maxInt(u64) value. Therefore, they wouldn't
+            /// be visible to queries with snapshot_latest (math.maxInt(u64 - 1).
+            fn exclude(range: *LevelKeyRange, exclude_range: KeyRange) void {
+                assert(range.non_empty());
+                var level = @fieldParentPtr(Self, "range", range);
+
+                if (compare_keys(exclude_range.key_max, range.key_max.?) == .eq) {
+                    if (level.iterator(
+                        .visible,
+                        &[1]u64{lsm.snapshot_latest},
+                        .descending,
+                        null,
+                    ).next()) |table| {
+                        range.key_max = table.key_max;
+                    } else {
+                        range.key_max = null;
+                    }
+                }
+                if (compare_keys(exclude_range.key_min, range.key_min.?) == .eq) {
+                    if (level.iterator(
+                        .visible,
+                        &[1]u64{lsm.snapshot_latest},
+                        .ascending,
+                        null,
+                    ).next()) |table| {
+                        range.key_min = table.key_min;
+                    } else {
+                        range.key_min = null;
+                    }
+                }
+                assert(range.empty() or range.non_empty());
+            }
+
+            fn include(range: *LevelKeyRange, include_range: KeyRange) void {
+                assert(range.empty() or range.non_empty());
+                if (range.key_min == null or compare_keys(
+                    include_range.key_min,
+                    range.key_min.?,
+                ) == .lt) {
+                    range.key_min = include_range.key_min;
+                }
+                if (range.key_max == null or compare_keys(
+                    include_range.key_max,
+                    range.key_max.?,
+                ) == .gt) {
+                    range.key_max = include_range.key_max;
+                }
+                assert(range.non_empty());
+            }
+
+            fn contains(range: *const LevelKeyRange, key: Key) bool {
+                assert(range.empty() or range.non_empty());
+                if (range.key_min == null and range.key_max == null) return false;
+                return compare_keys(key, range.key_min.?) != .lt and compare_keys(
```

Comment:
Popped into my head, but this could be made branchless with something like `boolToInt(not_lt) & boolToInt(not_gt)`. Not sure it would be worth it though.

Reply:
The `contains` function is branchless now (using a different strategy, though): https://github.com/tigerbeetle/tigerbeetle/blob/c2ee1b25726e53fe752f37bdadb21772a72d95f8/src/lsm/manifest_level.zig#L113-L116

Resolving this.